zcncore/transaction_query.go

```diff
@@ -0,0 +1,401 @@
+			return nil, nil, nil, err
+		}
+
+		// parse `confirmation` section as block header
+		confirmationBlockHeader, err = getBlockHeaderFromTransactionConfirmation(txnHash, confirmationBlock)
+		if err == nil {
+			return confirmationBlockHeader, confirmationBlock, nil, nil
+		}
+
+		Logger.Error("txn confirmation parse header error", err)
+
+		// parse `latest_finalized_block` section
+		lfbRaw, ok := confirmationBlock["latest_finalized_block"]
+		if !ok {
+			return confirmationBlockHeader, confirmationBlock, nil, err
+		}
+
+		err = json.Unmarshal([]byte(lfbRaw), &lfbBlockHeader)
+		if err == nil {
+			return confirmationBlockHeader, confirmationBlock, &lfbBlockHeader, ErrTransactionNotConfirmed
+		}
+
+		Logger.Error("round info parse error.", err)
+		return nil, nil, nil, err
+
+	}
+
+	return nil, nil, nil, thrown.Throw(ErrTransactionNotFound, strconv.Itoa(result.StatusCode))
+}
+
+func (tq *TransactionQuery) GetConsensusConfirmation(ctx context.Context, numSharders int, txnHash string) (*blockHeader, map[string]json.RawMessage, *blockHeader, error) {
+
+	maxConfirmation := int(0)
+	txnConfirmations := make(map[string]int)
+	var confirmationBlockHeader *blockHeader
+	var confirmationBlock map[string]json.RawMessage
+	var lfbBlockHeader *blockHeader
+	maxLfbBlockHeader := int(0)
+	lfbBlockHeaders := make(map[string]int)
+
+	// {host}/v1/transaction/get/confirmation?hash={txnHash}&content=lfb
+	err := tq.FromConsensus(ctx,
+		tq.buildUrl("", TXN_VERIFY_URL, txnHash, "&content=lfb"),
+		func(qr QueryResult) bool {
+			if qr.StatusCode != http.StatusOK {
+				return false
+			}
+
+			var cfmBlock map[string]json.RawMessage
+			err := json.Unmarshal([]byte(qr.Content), &cfmBlock)
+			if err != nil {
+				Logger.Error("txn confirmation parse error", err)
+				return false
+			}
+
+			// parse `confirmation` section as block header
+			cfmBlockHeader, err := getBlockHeaderFromTransactionConfirmation(txnHash, cfmBlock)
+			if err != nil {
+				Logger.Error("txn confirmation parse header error", err)
+
+				// parse `latest_finalized_block` section
+				if lfbRaw, ok := cfmBlock["latest_finalized_block"]; ok {
+					var lfb blockHeader
+					err := json.Unmarshal([]byte(lfbRaw), &lfb)
+					if err != nil {
+						Logger.Error("round info parse error.", err)
+						return false
+					}
+
+					lfbBlockHeaders[lfb.Hash]++
+					if lfbBlockHeaders[lfb.Hash] > maxLfbBlockHeader {
+						maxLfbBlockHeader = lfbBlockHeaders[lfb.Hash]
+						lfbBlockHeader = &lfb
+					}
+				}
+
+				return false
+			}
+
+			txnConfirmations[cfmBlockHeader.Hash]++
+			if txnConfirmations[cfmBlockHeader.Hash] > maxConfirmation {
+				maxConfirmation = txnConfirmations[cfmBlockHeader.Hash]
+
+				if maxConfirmation >= numSharders {
+					confirmationBlockHeader = cfmBlockHeader
+					confirmationBlock = cfmBlock
+
+					// it is consensus by enough sharders, and latest_finalized_block is valid
+					// return true to cancel other requests
+					return true
```

Comment:
Due to the FromConsensus() issue above, even we return true here, it won't stop sending requests

Reply:
It doesn't stop sending requests. it sends cancel signal for other requests.